### DISCUSSION TITLE
Form 和 Field 的 Effect Hooks 执行顺序是什么样的？

### DISCUSSION THREAD

**Discussion by dahui4dev** [alibaba/formily#1473] [Ideas]
## 场景

一个表单在修改内容之后希望自动提交表单。

## 问题

自动提交表单：最佳的操作是在哪个 Effects Hooks 中执行提交操作？

**Comment by janryWang** [comment]
监听onFormValuesChange就行了

**Reply by dahui4dev** [reply]
也就是说我在 onFormValuesChange 里直接 form.submit() 是吗？我现在是这么做的，如果我不调 submit() ，直接处理 onFormValuesChange 回调里的 form.values ，其实这个数据是没有 经过 validate 的，可能导致报错。

**Reply by janryWang** [reply]
最好做个节流吧，不然触发频率太高了

**Reply by dahui4dev** [reply]
嗯确实频率很高，所以来请教下，有没有最佳实践，怕我姿势不对。😄

**Reply by dahui4dev** [reply]
这里有个问题，即使我调用了 submit ，ArrayTable 行里的 input，依然没有触发，validate，需要修改一下 arrayTable 里的任意一个值才会触发检验。

![image](https://user-images.githubusercontent.com/6160798/119078147-370d2a80-ba28-11eb-9be9-8d6990b10b50.png)

就像这个一样。开始添加了新行是不会触发那个非空校验的。

![image](https://user-images.githubusercontent.com/6160798/119078172-442a1980-ba28-11eb-9398-048162c98d6a.png)


我下午补一个 codesandbox 代码

**Reply by janryWang** [reply]
点添加行本来就不应该触发内部的非空校验，除非你给array这个字段写一个自定义校验器

**Reply by dahui4dev** [reply]
看了下，自定义校验器能解决我的需求。这个高级功能一直没用过。挺好用的。

**Comment by dahui4dev** [comment]
@janryWang  有个想法：以下都是基于我目前需求考虑的，站在 formily 全局的角度不知道有没有用。

1. Form 这个 model 有没有可能提供一个 专门**校验**的方法。就类似于调用 submit 之后的那些校验流程，开放出来给用户单独调用校验但不提交结果。
2. Form 有没有可能统一设置一个 onBlur 的时候再触发 onFormValuesChange 的属性？这样降低触发 onFormValuesChange 的频率，或者说这个还是开放性的给用户自己处理。 如果用户自己处理，就目前我在什么时机处理比较合适？ 我在文档里找到（Field.onblur()这个是在什么情况下用的没太理解，我写的是 Schema 形式的）

**Reply by janryWang** [reply]
本来就有validate方法，文档看清楚点...

**Reply by janryWang** [reply]
第二点不用考虑了，太业务定制了，就像我说的，你用个debounce就行了

**Comment by janryWang** [comment]
对了，不要一直在一个话题上提其他问题，这非常不方便索引
